### PR TITLE
Upgrade Bglibpy to 4.0.27

### PR DESCRIPTION
**Pull request by wvangeit**

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-bglibpy/package.py
Comment (pramodk) [inline var/spack/repos/builtin/packages/py-bglibpy/package.py]: I see the point but in spack packaging convention, `import *` is done in all packages. @matz-e : opinion?

File: var/spack/repos/builtin/packages/py-bglibpy/package.py
Comment (pramodk) [inline var/spack/repos/builtin/packages/py-bglibpy/package.py]: Along with new as preferred, we prefer to keep few old version so that one can easily build older version.

File: var/spack/repos/builtin/packages/py-bglibpy/package.py
Comment (matz-e) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py]: I'd keep it in line with upstream.

File: var/spack/repos/builtin/packages/py-bglibpy/package.py
Comment (matz-e) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py]: IMO, `preferred` should only be set if an old version is preferred, and otherwise omitted. Makes version updates simpler, and spack automatically picks the newest version.

File: var/spack/repos/builtin/packages/py-bglibpy/package.py
Comment (wvangeit) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py]: Time for upstream to fix things ;-) Anyway, I've reverted to the upstream behavior

File: var/spack/repos/builtin/packages/py-bglibpy/package.py
Comment (wvangeit) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py]: Ok, I didn't add this preferred=True option myself, but I've removed it

File: var/spack/repos/builtin/packages/py-bglibpy/package.py
Comment (matz-e) [inline var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: I'd prefer if we could use tags rather than commits.

**Review by matz-e — DISMISSED** [review]

File: var/spack/repos/builtin/packages/py-bglibpy/package.py
Comment (wvangeit) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: There are no tags for these releases (in minor version releases get a tag)

**Comment by wvangeit** [comment]
A general comment, could you please heavily prioritise user-experience over review comments about details. This PR has now been in review for 27 days ... although the original commit would have worked perfectly. It's not very good advertisement for bbp spack if a simple version update of a package takes (at least) 27 days ...

File: var/spack/repos/builtin/packages/py-bglibpy/package.py
Comment (matz-e) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: Why not for patches? I don't like the idea that we just select a commit to be a release, tagging makes that "official".

**Comment by pramodk** [comment]
@wvangeit :  IIRC, my first review came in 2-3 days (and there was pause of 10 days, build was failing).

We (hpc) are reviewing asap but NSE, viz team should also have permissions to review and merge PRs. It's now collective responsibility.

But I understand and agree with you. We will keep an eye and try to get things asap in.

File: var/spack/repos/builtin/packages/py-bglibpy/package.py
Comment (wvangeit) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: That's how things work for bglibpy. I don't see why spack has to influence that.
And btw, I anyway consider a commit also much more stable. Afaik a tag can be moved, a commit uniquely identifies a state of the code.

Comment (matz-e) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: We had some software releases in `nix` that were "released" by assigning a commit a version in `nix`, and nothing else. To not have this arise in the future, I will insist on every software using tags.

**Review by pramodk — APPROVED** [review]

File: var/spack/repos/builtin/packages/py-bglibpy/package.py
Comment (mgeplf) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: I am confused as to what the 'resolution' was to this?

@wvangeit's point about tags being mutable (only need a `--force push`) does give me pause.  I hope no one would abuse this, but one never knows.

Comment (matz-e) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: True, just like people silently update release tarballs. It's frowned upon, and we should not fall for it. But you need to be pro-active to change it.

OTOH, it's really easy to forget documenting your release anywhere but in the packaging you do. Just run `master` do develop, and release periodically via `nix` is something I've seen in the short time I've been here.

**It's a lot easier to fall for the latter omission than to force the former**. Tags help overall to keep a record, even if they're not perfect. And it's something we can do centrally rather than peeking into every repository of software update we review, speaking of faster reviews.

Comment (wvangeit) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: In my case the version tracking is more or less automatic, by using: 
https://github.com/warner/python-versioneer
The patch number is just the number of commits since the last minor version tag

Comment (matz-e) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: It seems like `versioneer` does not support patch releases, and maybe you could tag minors instead? I see this `major.minor+git_commits` more as a development solution, not a deployment one.

Comment (mgeplf) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: > True, just like people silently update release tarballs. It's frowned upon, and we should not fall for it.

To me, it seems that the commit hash is the way to prevent this from happening.

I think I now understand where you're coming from - basically that you want to encourage people to release software more officially.  I think that's a great idea, and I appreciate it.  I believe that @jdcourcol is also hoping to have release/changelogs happen when software is released, so that 
ties in to the same thing.  I guess where we differ is that a *mandatory* tag be created  - maybe this can be something that needs to be officially agreed upon and documented somewhere.

Comment (matz-e) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: To me, tags and tarballs make things official. Just bumping a release number somewhere less so. Maybe there was a last minute hotfix (what would normally lead to the good-old fixup patch release), and the commit just one after the bump is then the official release — not documented anywhere? Given that we lost some of this "wild/free-style" release history by transitioning out of `nix`, I'm all for making things official. Create tags, and tarballs or wheels to match (that one could be optional). 

Writing a Changelog where you write down the commit hash that denotes the release (this would be properly documented IMO) will always be one commit behind the corresponding release, out of necessity.

Comment (mgeplf) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: > To me, tags and tarballs make things official.

I personally think that tags are good a idea, but it's not a universally held feeling, and I think it's fair that it's not a requirement.  The tarballs got uploaded to, so in that regard it was a release:
https://bbpteam.epfl.ch/repository/devpi/bbprelman/release/bglibpy/4.0.27

Should the spack package.py be pointing to our internal devpi?

Comment (matz-e) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: We can use those tarballs. I'll see how that works.

Comment (matz-e) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: Hm, I think I tried this already once. Not really trivial to get the URL right so that `spack checksum <my_package>` will just spit out the lines for new versions for you.

Comment (mgeplf) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: What do you mean trivial? Because it has a hash in the path?  Does the url need to exist anywhere but in the package.py, because then I'm ok w/ cut & paste.

Comment (matz-e) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: If you specify a "global" URL, Spack can replace the version information and re-use only one URL parameter.

I.e. for Black:
```
    homepage = "https://github.com/python/black"
    url      = "https://pypi.io/packages/source/b/black/black-18.9b0.tar.gz"

    version('19.3b0', sha256='68950ffd4d9169716bcb8719a56c07a2f4485354fec061cdd5910aa07369731c')
    version('18.9b0', sha256='e030a9a28f542debc08acceb273f228ac422798e5215ba2a791a6ddeaaca22a5')
```

Then you can just use `spack checksum black`, and it will give you additional `version` lines. Would be nice if I could get this to work with the BBP pypi, too.

Comment (matz-e) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: See also https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#url

Comment (mgeplf) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: Interesting.  So we'd lose the easily guessable URLs, but have a tarball.

I'd look at what devpi has to offer, but their documentation appears to be offline at the moment.

Comment (matz-e) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: Yes, noticed that, too.

Another possibility would be a `pip` source in Spack itself. More work, and potentially hard to get in upstream.

Comment (mgeplf) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: > Another possibility would be a pip source in Spack itself.

The setuptools/distutils ecosystem is already pretty complicated, I suspect it would add a lot of complexity to add pip.

BTW: has there been any talk of decoupling the spack code from the spack `spack/var/spack/repos/builtins`, because trying to search through the issues relating to the tool vs the packages isn't fun for me, and I assume it's way harder for the maintainers.

Comment (matz-e) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: > The setuptools/distutils ecosystem is already pretty complicated, I suspect it would add a lot of complexity to add pip.

May actually be pretty simple. I'm curious how far I could get given an afternoon to work on this.

> BTW: has there been any talk of decoupling the spack code from the spack spack/var/spack/repos/builtins, because trying to search through the issues relating to the tool vs the packages isn't fun for me, and I assume it's way harder for the maintainers.

Not that I'm aware of. @pramodk did you follow anything there?

Comment (pramodk) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: > Another possibility would be a pip source in Spack itself. More work, and potentially hard to get in upstream.
 
In general : using two package managers together is considered harmful !

Comment (pramodk) [reply var/spack/repos/builtin/packages/py-bglibpy/package.py:17]: > BTW: has there been any talk of decoupling the spack code from the spack spack/var/spack/repos/builtins, because trying to search through the issues relating to the tool vs the packages isn't fun for me, and I assume it's way harder for the maintainers.

It came to discussion few times, see [here](https://github.com/spack/spack/pull/8830). Once new upstream merge happen, we can start moving BBP packages in separate repo.

**Pull request merged**
